spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/BCECriterion.scala

```diff
@@ -37,48 +37,92 @@ import scala.reflect.ClassTag
  */
 @SerialVersionUID(- 1953992758534446600L)
 class BCECriterion[@specialized(Float, Double) T: ClassTag]
-(var weights: Tensor[T] = null, sizeAverage: Boolean = true)
+(val weights: Tensor[T] = null, sizeAverage: Boolean = true)
   (implicit ev: TensorNumeric[T]) extends TensorCriterion[T] {
   private val eps = 1e-12
-  if (weights != null) require(weights.dim() == 1,
-    "weights input should be 1-D Tensor" +
-    s"weights input dim(${weights.dim()})")
 
-  override def updateOutput(input: Tensor[T], target: Tensor[T]): T = {
-    require(input.nElement() == target.nElement())
+  val buffer: Tensor[T] = Tensor[T]
+
+  private var expendedWeights: Tensor[T] = null
 
-    if (null != weights && target.dim() != 1) {
-      weights = weights.view(1, target.size(2)).expandAs(target)
+  override def updateOutput(input: Tensor[T], target: Tensor[T]): T = {
+    require(input.size().sameElements(target.size()),
+      s"input size should be equal to target size, but got input size: ${input.size().toList}," +
+        s" target size: ${input.size().toList}")
+
+    if (weights != null) {
+      if (weights.nDimension() < input.nDimension()) {
+        require(weights.size().sameElements(input.size().tail),
+          s"weights size should be equal to input size or input size's tail, but got" +
+            s" input size: ${input.size().toList}, weights size: ${weights.size().toList}")
+      } else if (weights.nDimension() == input.nDimension()) {
+        require(weights.size().sameElements(input.size().tail),
```

Comment:
Is it `require(weights.size().sameElements(input.size())` ?

Reply:
yes, I'll fix that.